internal/object/objects.go

```diff
@@ -0,0 +1,114 @@
+package object
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icinga-go-library/com"
+	"github.com/icinga/icinga-go-library/database"
+	"github.com/icinga/icinga-go-library/types"
+	"github.com/icinga/icinga-notifications/internal/utils"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+	"sync"
+)
+
+var (
+	cache   = make(map[string]*Object)
+	cacheMu sync.Mutex
+)
+
+// DeleteFromCache deletes the Object from the global cache store matching the given ID (if any).
+func DeleteFromCache(id types.Binary) {
+	cacheMu.Lock()
+	defer cacheMu.Unlock()
+
+	delete(cache, id.String())
+}
+
+// RestoreMutedObjects restores all muted objects and their extra / ID tags from the database.
+// Note, this function only retrieves muted objects without non-recovered incident.
+//
+// Returns an error on any database failure and panics when trying to cache an object that's already in the cache store.
+func RestoreMutedObjects(ctx context.Context, db *database.DB) error {
+	objects := make(chan *Object)
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		defer close(objects)
+
+		clause := `WHERE mute_reason IS NOT NULL AND NOT EXISTS((SELECT 1 FROM incident WHERE object_id = object.id AND recovered_at IS NULL))`
+		query := fmt.Sprintf("%s %s", db.BuildSelectStmt(new(Object), new(Object)), clause)
+		err := utils.ExecAndApply[Object](ctx, db, query, nil, func(o *Object) {
+			o.db = db
+			o.Tags = map[string]string{}
+			o.ExtraTags = map[string]string{}
+
+			select {
+			case objects <- o:
+			case <-ctx.Done(): // The ctx error is going to be handled by utils.ExecAndApply.
+			}
+		})
+
+		return errors.Wrap(err, "cannot restore muted objects without an active incident")
+	})
+
+	g.Go(func() error {
+		bulks := com.Bulk(ctx, objects, db.Options.MaxPlaceholdersPerStatement, com.NeverSplit[*Object])
+
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case bulk, ok := <-bulks:
+				if !ok {
+					return nil
+				}
+
+				g.Go(func() error {
+					var ids []types.Binary
+					objectsMap := map[string]*Object{}
```

Comment:
The final size of these two variables is already known (`len(bulk)`), so they can be initialized to that capacity using `make()`.